packages/core/src/sagas/connectSaga.js

```diff
@@ -97,99 +84,175 @@ function rectifyUserID( directLine, userIDFromAction ) {
+function* reconnectSaga() {
+  for (;;) {
+    const { payload: { connectionStatus } } = yield take(UPDATE_CONNECTION_STATUS);
 
-    try {
-      const meta = {
-        userID: rectifyUserID( directLine, userIDFromAction ),
-        username
-      };
+    if (connectionStatus === ONLINE) {
+      break;
+    } else if (connectionStatus !== CONNECTING) {
+      throw new Error(`Failed to reconnect, DirectLineJS returned ${ connectionStatus }.`);
+    }
+  }
+}
 
-      let endDirectLine;
+// This is similar to behavior of redux-promise-middleware, but using saga instead of Promise.
+// We guarantee PENDING -> FULFILLING -> FULFILLED, or PENDING -> REJECTED. This will help us simplify logic in other part of code.
+// Note that after the saga is cancelled, subsequent call to put() will silently ignored.
+function* runAsyncEffect({ type, meta, payload }, callEffectFactory) {
+  try {
+    yield forkPut({ type: `${ type }_PENDING`, meta, payload });
 
-      yield put( { type: CONNECT_PENDING, meta } );
+    const result = yield callEffectFactory();
 
-      try {
-        endDirectLine = yield call( connectSaga, directLine );
-      } catch ( err ) {
-        yield put( { type: CONNECT_REJECTED, error: true, meta, payload: err } );
+    yield forkPut(
+      { type: `${ type }_FULFILLING`, meta, payload },
+      { type: `${ type }_FULFILLED`, meta, payload }
+    );
 
-        continue;
-      }
+    return result;
+  } catch (payload) {
+    yield forkPut({ type: `${ type }_REJECTED`, error: true, meta, payload });
 
-      // At this point, we established connection to Direct Line.
-      // If there are any errors from this point, we need to make sure we call endDirectLine() to release resources.
-      try {
-        yield put( { type: CONNECT_FULFILLING, meta, payload: { directLine } } );
-        yield put( { type: CONNECT_FULFILLED, meta, payload: { directLine } } );
+    throw payload;
+  }
+}
 
-        const terminateAction = yield take( [ DISCONNECT, negativeUpdateConnectionStatusAction ] );
+function* takeDisconnectAsError() {
+  yield take(DISCONNECT);
 
-        // Even if the connection is interrupted, we will still emit DISCONNECT_PENDING.
-        // This will makes handling logic easier. If CONNECT_FULFILLED, we guarantee DISCONNECT_PENDING.
-        yield put( { type: DISCONNECT_PENDING } );
+  throw new Error('disconnected');
+}
 
-        endDirectLine();
+function* runAsyncEffectUntilDisconnect(baseAction, callEffectFactory) {
+  // We cannot use saga cancel() here, because cancelling saga will prohibit us from sending *_REJECTED.
+  // Without REJECTED, it impact our assumptions around PENDING/FULFILLED/REJECTED.
+  return runAsyncEffect(
+    baseAction,
+    function* () {
+      const { result } = race({
+        _: takeDisconnectAsError(),
+        result: callEffectFactory()
+      });
+
+      return result;
+    }
+}
```

Comment:
this is missing an enclosing bracket

Reply:
Fixed. Sorry for committing bad code.

Now got unit tests to make sure everything work as expected.